File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] inDegree = new int[numCourses];
        Boolean[] visited = new Boolean[numCourses];
        ArrayList<Integer>[] adjMatrix = new ArrayList[numCourses];
        for(int i = 0; i < numCourses; i++) {
            adjMatrix[i] = new ArrayList<>();
            visited[i] = false;
        }

        for(int[] prerequisite: prerequisites) {
            int u = prerequisite[1];
            int v = prerequisite[0];
            adjMatrix[u].add(v);
            inDegree[v]++;
        }

        int[] courseOrder = new int[numCourses];
        int count = 0;
        Queue<Integer> q = new LinkedList<>();
        for(int i = 0; i < numCourses; i++) {
            if(inDegree[i] == 0) {
                visited[i] = true;
                courseOrder[count++] = i;
                q.add(i);
            }
        }
        
        while(!q.isEmpty()) {
            int u = q.poll();
            for(int v: adjMatrix[u]) {
                if(--inDegree[v] == 0) {
                    visited[v] = true;
                    courseOrder[count++] = v;
                    q.add(v);
                }
            }
        }

        if(count == numCourses) 
            return courseOrder;
        return new int[]{};
    }
}

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
from collections import defaultdict
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        inDegree = [0] * numCourses
        adjMap = defaultdict(list)

        for course, pre in prerequisites:
            inDegree[course] += 1
            adjMap[pre].append(course)
        
        res = []
        queue = deque(i for i in range(numCourses) if inDegree[i] == 0)
        while queue:
            curr = queue.popleft()
            res.append(curr)
            for neighbor in adjMap[curr]:
                inDegree[neighbor] -= 1
                if inDegree[neighbor] == 0:
                    queue.append(neighbor)
        return res if len(res) == numCourses else []